doc/source/administrator/optimization.md

```diff
@@ -383,3 +383,133 @@ Choosing the right _limit to guarantee ratio_ ratio is an art, not a science.
 For example, if the limit is 10GB, start with a guarantee of 5GB.
 Use a service such as Prometheus + Grafana to monitor the memory usage over time, and use this to decide whether to adjust your ratio.
+
+## Explicit memory and CPU allocated to core pods' containers
+
+The Helm chart creates several k8s Pods that typically have one container
+running in them, but sometimes more. In this section you are guided on how to
+explicitly specify how much CPU and memory they are guaranteed via _requests_
+and how much CPU and memory they are limited to via _limits_.
+
+To decide what requests and limits to set, some background knowledge is
+relevant:
+
+1. A [LimitRange](https://kubernetes.io/docs/concepts/policy/limit-range/)
+   resource in a namespace can provide defaults to containers without explicitly
+   set requests and limits.
+
+   Check the namespace where you deploy JupyterHub for such resource by doing
+   `kubectl get limitrange --namespace <k8s-namespace>`.
+
+1. Setting only resource limits _will make_ k8s assume the same in resource
+   requests, but setting only resource requests _will not make_ k8s assume any
+   resource limits.
+
+1. Setting resource requests improves the reliability of the container's
+   execution. Setting resource limits improves the consistency of the execution
+   of the specific container and other containers on the nodes.
+
+1. Requested memory is reserved and unavailable for other containers on a node
+   to use, but any requested CPU can be used by other containers until its
+   required by the requesting container.
+
+1. A container running out of memory is typically terminated and restarted as a
+   process is killed by a _Out Of Memory Killer_ (OOMKiller). When this happens
+   you should see a trace of it by using `kubectl describe pod --namespace <k8s-namespace> <k8s-pod-name>`
+   and `kubectl logs --previous --namespace <k8s-namespace> <k8s-pod-name>`.
+
+1. A container starved of CPU could act up in many unique ways and is harder to
+   debug. Various timeouts can be clues to suspect CPU starvation.
+
+1. When scheduling a Pod on a node, the [_effective
+   requests/limits_](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#resources)
+   are considered. As a Pod's init containers are run in sequence before the
+   Pod's main containers are started, the effective requests/limits are
+   calculated as the highest of the init containers requests/limits and the sum
+   of the main containers requests/limits.
+
+1. Software may not be able to make use of more than 1 CPU because it doesn't
+   support running code concurrently across multiple CPU cores. The `hub` pod
+   running JupyterHub in Python and the `proxy` pod running
+   ConfigurableHTTPProxy in NodeJS are such applications.
+
+Some additional more technical details are:
+
+1. A container requesting `0` CPU will still be granted a share of `0.002` CPU,
+   or `2m` CPU, it is a lower limit.
```

Comment:
Yes I know =/ Well, i think it is out of scope for k8s to declare it, but the docker Container Runtime Interface (CRI) at least does this. I considered this as well and decided that it was better to not try to make a disclaimer about that it may not be true with different CRI or different configurations of that CRI etc.

I think its a relevant point to clarify that one can expect 0 0 requests to actually be granted some minimal share of a CPU. If you would schedule something to run with absolute zero CPU it wouldn't make sense at all anyhow so I think its safe to guess all CRI will provide some minimum.

Reply:
I updated "container engine" to "container runtime".